src/OpenTelemetry/Internal/InterlockedHelper.cs

```diff
@@ -0,0 +1,58 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Runtime.CompilerServices;
+
+namespace OpenTelemetry.Internal;
+
+internal static class InterlockedHelper
+{
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static void Add(ref double location, double value)
+    {
+        // Note: Not calling InterlockedHelper.Read here on purpose because it
+        // is too expensive for fast/happy-path. If the first attempt fails
+        // we'll end up in an Interlocked.CompareExchange loop anyway.
+        double currentValue = Volatile.Read(ref location);
+
+        double updatedValue;
+        unchecked
+        {
+            updatedValue = currentValue + value;
+        }
+
+        var returnedValue = Interlocked.CompareExchange(ref location, updatedValue, currentValue);
+        if (returnedValue != currentValue)
+        {
+            AddRare(ref location, value, returnedValue);
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static double Read(ref double location)
+        => Interlocked.CompareExchange(ref location, double.NaN, double.NaN);
```

Comment:
Why do you need this to be a CompareExchange rather than just Volatile.Read? You need a full fence for some reason?

Reply:
The code before was...

```csharp
this.snapshotValue.AsDouble = Interlocked.CompareExchange(ref this.runningValue.AsDouble, 0.0, double.NegativeInfinity);
```

Now it is calling this helper...

```csharp
this.snapshotValue.AsDouble = InterlockedHelper.Read(ref this.runningValue.AsDouble);
```

What that code _needs_ is no stale reads. Does `Interlocked.CompareExchange` (full fence) give us 100% guaranteed up-to-date values? My understanding with `Volatile.Read` is it _could_ give us back something stale.

Reply:
> What that code needs is no stale reads.

Define stale. Volatile.Read will ensure it's not just using some enregistered value. But beyond that, I don't see how CompareExchange(ref location, sentinel, sentinel) would provide strong guarantees about "up-to-date". The exact moment after you read the value but before you do anything with it it could be updated concurrently, at which point the value isn't "up to date".

Reply:
How OpenTelemetry SDK is working is many threads are recording values. For this double case it is either of these:

```csharp
            case AggregationType.DoubleSumIncomingDelta:
                {
                    InterlockedHelper.Add(ref this.runningValue.AsDouble, number);
                    break;
                }

            case AggregationType.DoubleSumIncomingCumulative:
                {
                    Interlocked.Exchange(ref this.runningValue.AsDouble, number);
                    break;
                }
```

There is a dedicated thread which does a collection operation periodically. Its job is to grab a snapshot of the current/running state. That is this block:

```csharp
            case AggregationType.DoubleSumIncomingDelta:
            case AggregationType.DoubleSumIncomingCumulative:
                {
                    if (outputDelta)
                    {
                        double initValue = InterlockedHelper.Read(ref this.runningValue.AsDouble);
                        this.snapshotValue.AsDouble = initValue - this.deltaLastValue.AsDouble;
                        this.deltaLastValue.AsDouble = initValue;
                        this.MetricPointStatus = MetricPointStatus.NoCollectPending;

                        // Check again if value got updated, if yes reset status.
                        // This ensures no Updates get Lost.
                        if (initValue != InterlockedHelper.Read(ref this.runningValue.AsDouble))
                        {
                            this.MetricPointStatus = MetricPointStatus.CollectPending;
                        }
                    }
                    else
                    {
                        this.snapshotValue.AsDouble = InterlockedHelper.Read(ref this.runningValue.AsDouble);
                    }

                    break;
                }
```

The delta case needs to compute the change since the last collect:

```csharp
                        double initValue = InterlockedHelper.Read(ref this.runningValue.AsDouble);
                        this.snapshotValue.AsDouble = initValue - this.deltaLastValue.AsDouble;
                        this.deltaLastValue.AsDouble = initValue;
```

It's OK if the value changes right before or right after but we need to make sure whatever we got was true and not some cached or local thing. Does that help?

Reply:
> It's OK if the value changes right before or right after but we need to make sure whatever we got was true and not some cached or local thing.

I still don't understand. With a CompareExchange, you might read the value, then the thread might be context switched out, and 10 minutes later it might start running again and you continue with the 10 minute old value. What's the difference between that and "some cached" thing?

Reply:
> you might read the value, then the thread might be context switched out, and 10 minutes later it might start running again and you continue with the 10 minute old value. What's the difference between that and "some cached" thing?

CPU's context switching is not under our control, but we still want to put our best effort in capturing as many updates as possible (that happened before we run `Collect`). That's we would like to avoid reading "some cached" thing.

Reply:
This PR appears to be trying to minimize overheads for things like a non-inlined method call. If that level of overhead is a concern, you're spending way more cycles than that using Interlocked.CompareExchange instead of just using Volatile.Read.  Did anyone ever confirm the current approach is indeed "capturing [more] updates] with CompareExchange? I suspect it wasn't.

Reply:
Even if it does improve how many are captured, you also need to think about the impact the additional synchronization has on the code doing the updates.  Presumably you care about the throughput of those, yes? Using an Interlocked.CompareExchange here may give you a value more synchronized with the other uses of those variables, but that's because it's actually synchronizing, and that's not just an overhead for this reader, but also an overhead for those other uses.

For example, here's a quick and dirty microbenchmark on the cost of those Interlocked.Adds:
```C#
using System.Diagnostics;

long value = 0;
var sw = new Stopwatch();
Parallel.Invoke(
    () =>
    {
        while (true)
        {
            sw.Restart();
            for (int i = 0; i < 100_000_000; i++)
            {
                Interlocked.Add(ref value, 1);
            }
            Console.WriteLine(sw.Elapsed);
        }
    },
    () =>
    {
        while (true) Interlocked.CompareExchange(ref value, 0, 0);
        //while (true) Volatile.Read(ref value);
    });
```
When I run that, I see it taking ~2.0s to ~2.3s per 100M updates. If I then switch the CompareExchange to be a Read, that drops to between ~1.6s and ~1.9s. The "up-to-date"ness is achieved by locking the cache line, and you're paying for that in both places.

Reply:
> The "up-to-date"ness is achieved by locking the cache line, and you're paying for that in both places

Our use-case has one major difference from the setup that you shared. The collect thread runs periodically usually like every 30 sec or 1 minute so the frequency of CompareExchange calls made from the collect thread is not that high. I believe that should greatly reduce the perf impact.

From a business standpoint, this code is used for collecting metrics. It's very common for users to have their alerts configured based on metrics. This is our main motivation behind doing our best to capture as many updates as possible before running collect. Any updates that we miss, would only be reported in the next collect cycle (which would be like a minute later).

Reply:
@stephentoub 

Here is our version for `long` measurements:

```csharp
            case AggregationType.LongSumIncomingDelta:
            case AggregationType.LongSumIncomingCumulative:
                {
                    if (outputDelta)
                    {
                        long initValue = Interlocked.Read(ref this.runningValue.AsLong);
                        this.snapshotValue.AsLong = initValue - this.deltaLastValue.AsLong;
                        this.deltaLastValue.AsLong = initValue;
                        this.MetricPointStatus = MetricPointStatus.NoCollectPending;

                        // Check again if value got updated, if yes reset status.
                        // This ensures no Updates get Lost.
                        if (initValue != Interlocked.Read(ref this.runningValue.AsLong))
                        {
                            this.MetricPointStatus = MetricPointStatus.CollectPending;
                        }
                    }
                    else
                    {
                        this.snapshotValue.AsLong = Interlocked.Read(ref this.runningValue.AsLong);
                    }

                    break;
                }
```

Also uses `Interlocked.Read`, do you think it would be better to use `Volatile.Read` there too?

Reply:
> This is our main motivation behind doing our best to capture as many updates as possible before running collect. Any updates that we miss, would only be reported in the next collect cycle (which would be like a minute later).

Can you share the data you have that demonstrates using CompareExchange is actually improving this?

Reply:
> Also uses Interlocked.Read, do you think it would be better to use Volatile.Read there too?

`Interlocked.Read(ref value)` is just `CompareExchange(ref value, 0, 0)`, so it's effectively the same thing you're doing with the doubles. Personally, I'd just use `Volatile.Read`. I'm really interested in seeing the data that suggests there's any meaningful benefit to using CompareExchange instead in these scenarios; otherwise, I'd stick with the idiomatic answer that's also cheaper.

Reply:
> Can you share the data you have that demonstrates using CompareExchange is actually improving this?

We haven't tested this, and we most likely would not anytime soon. It would require elaborate testing setups with different kinds of hardware. That might not be the best use of our time. Ideally, we would like to rely on the official docs for this kind of information instead of running experiments ourselves and gathering this kind of data.

In this particular case, our understanding/assumption is in line with what you mentioned in one of your previous comments:
> Using an Interlocked.CompareExchange here may give you a value more synchronized with the other uses of those variables, but that's because it's actually synchronizing,

Reply:
My guess is you'd find it makes essentially zero difference, other than slowing down the writing code.